[MAINT] Fix `github_organization_ruleset` and `github_repository_ruleset` with `push` target

deiga opened the pull request:
<!-- Please refer to our contributing docs for any questions on submitting a pull request -->
<!-- Issues are required for both bug fixes and features. -->
Resolves #2929, #2467

----

### Before the change?
<!-- Please describe the current behavior that you are modifying. -->

* The provider would crash from a 422 error response when following the example in the docs
* Removing `ref_name` would cause the provider to Panic as `ref_name` is a required field
* One could add `rules` which weren't valid for `push` rulesets

### After the change?
<!-- Please describe the behavior or changes that are being added by this PR. -->

* The provider should correctly apply `push` rulesets to an organization
* `ref_name` should no longer be needed to be set for `push` target
* `conditions` & `target` validation logic should ensure correct fields are populated

### Pull request checklist
- [ ] ~Schema migrations have been created if needed ([example](https://github.com/integrations/terraform-provider-github/pull/2820/files))~
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Does this introduce a breaking change?
<!-- If this introduces a breaking change make sure to note it here any what the impact might be -->

Please see our docs on [breaking changes](https://github.com/octokit/.github/blob/master/community/breaking_changes.md) to help!

- [ ] Yes
- [x] No

----

deiga committed:
Update descriptions and validations

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add `CustomizeDiff` logic to validate on `plan`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add first validation test

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add validation error when `conditions` is missing

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add further validation tests

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Remove unnecessary skip blocks as `individual` and `anonymous` access to org rulesets will never be a thing

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Switch `ref_name` to `Optional` as `push` doesn't need a `ref_name`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

nickfloyd commented on github/resource_github_organization_ruleset.go:
```diff
@@ -713,3 +714,50 @@ func resourceGithubOrganizationRulesetImport(d *schema.ResourceData, meta any) (
+	conditions := d.Get("conditions").([]any)[0].(map[string]any)
+	if conditions["ref_name"] == nil || conditions["repository_name"] == nil || conditions["repository_id"] == nil {
+		return fmt.Errorf("ref_name and repository_name or ref_name and repository_id must be set for branch and tag targets")
+	}
```
This should be simplified to make what is required a bit more clear.

```suggestion
    if conditions["ref_name"] == nil {
        return fmt.Errorf("ref_name must be set for branch and tag targets")
    }
    if conditions["repository_name"] == nil && conditions["repository_id"] == nil {
        return fmt.Errorf("repository_name or repository_id must be set for branch and tag targets")
    }
```

nickfloyd commented on github/resource_github_organization_ruleset.go:
```diff
@@ -86,12 +88,12 @@ func resourceGithubOrganizationRuleset() *schema.Resource {
 						"ref_name": {
 							Type:     schema.TypeList,
-							Required: true,
+							Optional: true,
```
This is a non breaking change ➕

deiga committed:
Add Debug logging with `tflog` to validation

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix validation as `ref_name`, `repository_name` and `repository_id` are empty lists by default

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Remove unnecessary panic test

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Improve validation output messages

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell reviewed (commented):
I think this change likely wants to wait for the SDK upgrade as a lot of this area is modified in future versions.

FYI the error behaviours previously seen should have been mitigated by #2705 so if there is an error the provider should handle it gracefully.

deiga committed:
Fix condition to require only one of `repository_name` or `repository_id`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -35,8 +37,8 @@ func resourceGithubOrganizationRuleset() *schema.Resource {
-				ValidateFunc: validation.StringInSlice([]string{"branch", "tag", "push"}, false),
-				Description:  "Possible values are `branch`, `tag` and `push`. Note: The `push` target is in beta and is subject to change.",
+				ValidateFunc: validation.StringInSlice([]string{"branch", "tag", "push", "repository"}, false),
+				Description:  "The target of the ruleset. Possible values are `branch`, `tag`, `push` and `repository`.",
```
I literally just added support for them in https://github.com/google/go-github/pull/3850 and they've just been released in [v80.0.0](https://github.com/google/go-github/releases/tag/v80.0.0).

stevehipwell replied:
This needs removing.

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -86,12 +88,12 @@ func resourceGithubOrganizationRuleset() *schema.Resource {
 				Optional:    true,
 				MaxItems:    1,
-				Description: "Parameters for an organization ruleset condition. `ref_name` is required alongside one of `repository_name` or `repository_id`.",
+				Description: "Parameters for an organization ruleset condition. For `branch` and `tag` targets, `ref_name` is required alongside one of `repository_name` or `repository_id`. The `push` target does not require `ref_name` conditions. For `repository` target, the conditions object should only contain the `repository_name` and the `repository_id`.",
```
The wording isn't right even if the field names are, the REST API makes a complete mess of this too. My point was that it'd be worth making the comment clearer as to what the fields actually do.

deiga replied:
You are right that we can try to describe it better than the API docs. I'll try my hand at that.

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -659,46 +662,46 @@ func resourceGithubOrganizationRulesetRead(d *schema.ResourceData, meta any) err
 
 	var ruleset *github.RepositoryRuleset
 
 	ruleset, _, err = client.Organizations.UpdateRepositoryRuleset(ctx, org, rulesetID, *rulesetReq)
```
Checking if the bypass actors have changed and the length is `0` here would allow you to use the clear function to resolve #2952. See https://github.com/integrations/terraform-provider-github/pull/2891#discussion_r2591993478.

deiga replied:
@stevehipwell The problem here is, that go-github doesn't have `UpdateRulesetClearBypassActor` for Organization rulesets :/

deiga committed:
Add validation to `required_workflow.path`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Rename test resources for easier debugging

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix linter issues

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add validation to ensure `rules.required_status_checks.required_checks.context` is not empty

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add test to ensure that `required_checks` is always required

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell reviewed (changes requested):
Would it be possible to either de-scope this PR or open a new PR with the smallest number of changes possible to fix the outstanding bugs?

deiga commented on github/config.go:
```diff
@@ -47,7 +47,7 @@ var GHECDataResidencyHostMatch = regexp.MustCompile(`^[a-zA-Z0-9.\-]+\.ghe\.com\
 	client.Transport = NewEtagTransport(client.Transport)
 	client.Transport = NewRateLimitTransport(client.Transport, WithWriteDelay(writeDelay), WithReadDelay(readDelay), WithParallelRequests(parallelRequests))
-	client.Transport = logging.NewSubsystemLoggingHTTPTransport("GitHub", client.Transport)
+	client.Transport = logging.NewLoggingHTTPTransport(client.Transport)
```
We don't inititialize the `GitHub` subsystem anywhere and from my understanding subsystems are usually per-resource things.
With an unitialized Subsystem the logs get spammed with warnings about it.

But this implementation is "actually" in this PR https://github.com/integrations/terraform-provider-github/pull/2976 and any modifications to it should happen there

deiga commented on github/resource_github_organization_ruleset.go:
```diff
@@ -45,7 +50,7 @@ func resourceGithubOrganizationRuleset() *schema.Resource {
 			},
 			"bypass_actors": {
-				Type:             schema.TypeList,
+				Type:             schema.TypeList, // TODO: These are returned from GH API sorted by actor_id, we might want to investigate if we want to include sorting
```
That sounds sensible, need to investigate where the best place for the sorting is

deiga committed:
Fix tests after rebase

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Improve legibility of `conditions` description

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Update descriptions

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add Acc test for push ruleset.

This turns out to be failing as there is a bug in our implementation! Unit tests and fix coming up

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add failing test for `flattenConditions` with no `ref_name` condition

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix `flattenConditions` to work with `push` rulesets

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add more tests for `flattenConditions`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Enable debug logging in `flattenConditions`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Ensures that `flattenConditions` returns an empty list on empty API response

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add validation for `push` `rules`

As they differ from `branch` and `tag` rules

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Get `TestAccGithubRepositoryRulesets` to work

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
`repository_ruleset`: Add tests for validations

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
`repository_ruleset`: Implement validations for `target`, `conditions` and `rules`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Updated ruleset docs

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Extract validation functions to separate utils file with unit tests

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix push ruleset test config

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Remove `repository` target after thorough testing that it doesn't do anything

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell commented on github/repository_rules_validation_utils_test.go:
```diff
@@ -0,0 +1,225 @@
+	"testing"
+)
+
+func TestValidateConditionsFieldForPushTarget(t *testing.T) {
```
```suggestion
func Test_validateConditionsFieldForPushTarget(t *testing.T) {
```

This would be the idiomatic name for an unexported method test.

deiga replied:
Oh! Thanks 🙏

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -831,3 +844,37 @@ func resourceGithubOrganizationRulesetImport(ctx context.Context, d *schema.Reso
 	return []*schema.ResourceData{d}, nil
 }
+
+func validateConditionsFieldBasedOnTarget(ctx context.Context, d *schema.ResourceDiff, meta any) error {
```
If this is scoped to the organization ruleset then the name should reflect that, otherwise it should probably be in the _util\_rules.go_ file (idiomatically I'd expect this to be called _util\_ruleset.go_).

deiga replied:
Ugh, right. I keep forgetting that all functions within a package are visible 😬

stevehipwell reviewed (changes requested):
I've added some review comments, mainly about the code structure.

deiga committed:
Use idiomatic test naming convention

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Address code structure comment in tests

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix inconsistent logging

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Refactor to use typed constant string for ruleset `Target`

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Fix indentation issue with `github_repository_file` and heredocs

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Rename files to be more sensible

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Refactor validation functions so that Repo and Org share almost everything

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell commented on github/resource_github_repository.go:
```diff
@@ -1019,7 +1019,7 @@ func resourceGithubRepositoryUpdate(ctx context.Context, d *schema.ResourceData,
 	}
 
-	if d.HasChanges("visibility", "private") {
+	if d.HasChanges("visibility", "private") && !d.Get("archived").(bool) {
```
This should be in a separate PR, and I'm also not sure if it's valid to disable changing the visibility level when archiving (the execution order issue is a separate concern).

stevehipwell commented on github/resource_github_repository_test.go:
```diff
@@ -1315,19 +1315,19 @@ func Test_expandPages(t *testing.T) {
-	resource := resourceGithubRepository()
-	schema := resource.Schema["topics"].Elem.(*schema.Schema)
-	diags := schema.ValidateDiagFunc("ef69e1a3-66be-40ca-bb62-4f36186aa292", cty.Path{cty.GetAttrStep{Name: "topic"}})
+	repositoryResourceSchema := resourceGithubRepository()
```
Shouldn't these changes be in a separate PR?

stevehipwell commented on github/resource_github_team_repository.go:
```diff
@@ -138,7 +138,7 @@ func resourceGithubTeamRepositoryRead(d *schema.ResourceData, meta any) error {
 			}
 		}
-		return err
+		return repoErr
```
Shouldn't this fix be in a separate PR?

stevehipwell reviewed (changes requested):
@deiga I've added some comments but the main concern here is that we keep only the relevant changes in the PR. If you could tidy the PR up, rebase, and then I'll give this a full PR with the aim of getting it merged ASAP.

deiga committed:
Address review comments

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell reviewed (changes requested):
This is looking good. I've added a few more comments, if you could address these then I think we're ready to review the new validation logic.

deiga committed:
Use `github.RepositoryRuleType` instead of strings

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Refactor `flattenRules` and `expandRules` to use context

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga commented:
@stevehipwell Thanks for the comments! I addressed them all now :)

deiga committed:
Fix failing validation

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -488,9 +495,10 @@ func resourceGithubOrganizationRuleset() *schema.Resource {
-													Description: "The path to the workflow YAML definition file.",
+													Type:             schema.TypeString,
+													Required:         true,
+													ValidateDiagFunc: toDiagFunc(validation.StringMatch(regexp.MustCompile(`^\.github\/workflows\/.*$`), "Path must be in the .github/workflows directory"), "path"),
```
```suggestion
													ValidateDiagFunc: validation.ToDiagFunc(validation.StringMatch(regexp.MustCompile(`^\.github\/workflows\/.*$`), "Path must be in the .github/workflows directory")),
```
As an example of a change from the comment above.

stevehipwell reviewed (changes requested):
@deiga this is looking good, just a couple of review comments to look at.

deiga committed:
Refactor to use `ValidateDiagFunc` and `validation.ToDiagFunc` in repo ruleset

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Use `strings.Join` to make `Description` of `target` dynamic

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add validation to `operator` attributes

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Add missing validations to repo ruleset attributes

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Refactor to use `ValidateDiagFunc` and `validation.ToDiagFunc` in org ruleset

deiga committed:
Use `strings.Join` to make `Description` of `target` dynamic

deiga committed:
Add validation to `operator` attributes

deiga committed:
Add missing validations to org ruleset attributes

deiga committed:
Ensure attribute names map correctly to rule names

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

deiga committed:
Don't ignore errors

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell commented on github/resource_github_organization_ruleset.go:
```diff
@@ -651,10 +663,18 @@ func resourceGithubOrganizationRulesetCreate(ctx context.Context, d *schema.Reso
+		return diag.FromErr(err)
+	}
 
 	tflog.Info(ctx, fmt.Sprintf("Created organization ruleset: %s/%s (ID: %d)", owner, name, *ruleset.ID), map[string]any{
```
ruleset.GetID()

stevehipwell reviewed (changes requested):
This PR is looking great @deiga; I think we're down to the last minor consistency issue.

deiga committed:
Address review comments

Signed-off-by: Timo Sand <timo.sand@f-secure.com>

stevehipwell reviewed (approved):
LGTM
